File: src/main/java/waterpower/common/block/BlockRotor.java
package waterpower.common.block;

import javax.annotation.Nullable;

import net.minecraft.block.material.Material;
import net.minecraft.block.state.IBlockState;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.item.ItemBlock;
import net.minecraft.item.ItemStack;
import net.minecraft.tileentity.TileEntity;
import net.minecraft.util.math.BlockPos;
import net.minecraft.util.EnumFacing;
import net.minecraft.util.EnumHand;
import net.minecraft.world.World;
import waterpower.common.block.tileentity.TileEntityRotor;
import waterpower.common.item.rotor.ItemRotor;

public abstract class BlockRotor extends BlockWaterPower {

    public BlockRotor(String id, Material material, Class<? extends ItemBlock> c) {
        super(id, material, c);
    }

    @Override
    public boolean onBlockActivated(World world, BlockPos pos, IBlockState state, EntityPlayer entityPlayer, EnumHand hand, @Nullable ItemStack heldItem, EnumFacing side, float hitX, float hitY, float hitZ) {
        TileEntity te = world.getTileEntity(pos);
        if (te instanceof TileEntityRotor) {
            TileEntityRotor ter = (TileEntityRotor) te;
            ItemStack now = ter.slotRotor.get();
            if (heldItem != null && heldItem.getItem() instanceof ItemRotor) {
                ter.slotRotor.put(heldItem);
                int id = entityPlayer.inventory.currentItem;
                if (0 <= id && id < 9)
                    entityPlayer.inventory.mainInventory[id] = null;
                if (now != null)
                    entityPlayer.inventory.addItemStackToInventory(now);
                return true;
            } else if (heldItem == null && entityPlayer.isSneaking()) {
                ter.slotRotor.put(null);
                if (now != null)
                    entityPlayer.inventory.addItemStackToInventory(now);
            }
        }
        return super.onBlockActivated(world, pos, state, entityPlayer, hand, heldItem, side, hitX, hitY, hitZ);
    }
}

File: src/main/kotlin/waterpower/common/block/BlockRotor.kt
/**
 * Copyright (c) Huang Yuhui, 2017
 *
 * "WaterPower" is distributed under the terms of the Minecraft Mod Public
 * License 1.0, or MMPL. Please check the contents of the license located in
 * http://www.mod-buildcraft.com/MMPL-1.0.txt
 */
package waterpower.common.block

import net.minecraft.block.ITileEntityProvider
import net.minecraft.block.material.Material
import net.minecraft.block.state.IBlockState
import net.minecraft.entity.player.EntityPlayer
import net.minecraft.item.ItemStack
import net.minecraft.util.EnumFacing
import net.minecraft.util.EnumHand
import net.minecraft.util.math.BlockPos
import net.minecraft.world.World
import waterpower.common.block.tile.TileEntityRotorGenerator
import waterpower.common.item.ItemRotor
import waterpower.util.emptyStack
import waterpower.util.isStackEmpty

abstract class BlockRotor<T>(id: String, material: Material, typeClass: Class<T>, types: Array<T>)
    : BlockEnumTile<T>(id, material, typeClass, types)
where T : Enum<T>, T : waterpower.common.INameable, T : ITileEntityProvider {

    override fun onBlockActivated(worldIn: World, pos: BlockPos, state: IBlockState, playerIn: EntityPlayer, hand: EnumHand, stack: ItemStack?, facing: EnumFacing, hitX: Float, hitY: Float, hitZ: Float): Boolean {
        val te = worldIn.getTileEntity(pos)
        if (te is TileEntityRotorGenerator) {
            val now = te.slotRotor.get()
            val heldItem = playerIn.getHeldItem(hand)
            if (heldItem != null && heldItem.getItem() is ItemRotor) {
                te.slotRotor.put(heldItem)
                val id = playerIn.inventory.currentItem
                if (id in 0 until 9)
                    playerIn.inventory.mainInventory[id] = emptyStack
                if (!isStackEmpty(now))
                    playerIn.inventory.addItemStackToInventory(now)
                return true
            } else if (heldItem == null && playerIn.isSneaking()) {
                te.slotRotor.put(emptyStack)
                if (!isStackEmpty(now))
                    playerIn.inventory.addItemStackToInventory(now)
            }
        }
        return super.onBlockActivated(worldIn, pos, state, playerIn, hand, stack, facing, hitX, hitY, hitZ)
    }
}